jlm/llvm/ir/attribute.hpp

```diff
@@ -267,181 +220,148 @@ class int_attribute final : public enum_attribute
   {}
 
-  const jlm::rvsdg::valuetype &
+  [[nodiscard]] const jlm::rvsdg::valuetype &
   type() const noexcept
   {
     return *type_;
   }
 
-  virtual bool
+  bool
   operator==(const attribute &) const override;
 
-  virtual std::unique_ptr<attribute>
-  copy() const override;
-
-  static std::unique_ptr<attribute>
-  create_byval(std::shared_ptr<const jlm::rvsdg::valuetype> type)
-  {
-    return std::make_unique<type_attribute>(kind::ByVal, std::move(type));
-  }
-
-  static std::unique_ptr<attribute>
-  CreateStructRetAttribute(std::shared_ptr<const jlm::rvsdg::valuetype> type)
-  {
-    return std::make_unique<type_attribute>(kind::StructRet, std::move(type));
-  }
-
 private:
   std::shared_ptr<const jlm::rvsdg::valuetype> type_;
 };
 
-/** \brief Attribute set
- */
-class attributeset final
-{
-  class constiterator;
-
-public:
-  ~attributeset()
-  {}
-
-  attributeset() = default;
-
-  attributeset(std::vector<std::unique_ptr<attribute>> attributes)
-      : attributes_(std::move(attributes))
-  {}
-
-  attributeset(const attributeset & other)
-  {
-    *this = other;
-  }
-
-  attributeset(attributeset && other)
-      : attributes_(std::move(other.attributes_))
-  {}
+}
 
-  attributeset &
-  operator=(const attributeset & other);
+namespace jlm::util
+{
 
-  attributeset &
-  operator=(attributeset && other)
+template<>
+struct Hash<jlm::llvm::enum_attribute>
+{
+  std::size_t
+  operator()(const jlm::llvm::enum_attribute & attribute) const noexcept
   {
-    if (this == &other)
-      return *this;
-
-    attributes_ = std::move(other.attributes_);
-
-    return *this;
+    return std::hash<jlm::llvm::attribute::kind>()(attribute.kind());
   }
+};
 
-  constiterator
-  begin() const;
-
-  constiterator
-  end() const;
-
-  void
-  insert(const attribute & a)
+template<>
+struct Hash<jlm::llvm::int_attribute>
+{
+  std::size_t
+  operator()(const jlm::llvm::int_attribute & attribute) const noexcept
   {
-    attributes_.push_back(a.copy());
+    return std::hash<jlm::llvm::attribute::kind>()(attribute.kind())
+         ^ std::hash<uint64_t>()(attribute.value()) << 1;
   }
+};
 
-  void
-  insert(std::unique_ptr<attribute> a)
+template<>
+struct Hash<jlm::llvm::string_attribute>
+{
+  std::size_t
+  operator()(const jlm::llvm::string_attribute & attribute) const noexcept
   {
-    attributes_.push_back(std::move(a));
+    return std::hash<std::string>()(attribute.kind())
+         ^ std::hash<std::string>()(attribute.value()) << 1;
   }
+};
 
-  bool
-  operator==(const attributeset & other) const noexcept
+template<>
+struct Hash<jlm::llvm::type_attribute>
+{
+  std::size_t
+  operator()(const jlm::llvm::type_attribute & attribute) const noexcept
   {
-    /*
-      FIXME: Ah, since this is not a real set, we cannot cheaply implement a comparison.
-    */
-    return false;
+    // FIXME: I have no better idea of how to handle this currently
+    return std::hash<ptrdiff_t>()(reinterpret_cast<ptrdiff_t>(&attribute));
```

Comment:
@haved @caleridas Any better idea?

Reply:
This might be problematic as two `type_attribute` can compare equal with `operator==` yet have different hashes.

Would hashing the kind and type and xoring them together work?

Reply:
@phate I checked the definition of std::hash, and they require == equality to imply identical hashes

Reply:
Yeah, this won't work properly unless we implement hashing capabilities for `rvsdg::type`. I will take a stab at it in a new PR, then we can merge this one.

Reply:
Fixed.